File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/main/hardware/DriveTrain.java
package org.firstinspires.ftc.teamcode.main.hardware;

import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorSimple;

import org.openftc.revextensions2.ExpansionHubMotor;


public class DriveTrain extends Hardware {

    // hahaha this setup is so fucked i just realized now that the motor array and the motorpower array dont match up
    public static double movementX, movementY, movementTurn;
    private final ExpansionHubMotor frontLeft, frontRight, backLeft, backRight;
    private final ExpansionHubMotor[] motors;

    public DriveTrain(ExpansionHubMotor FL, ExpansionHubMotor FR, ExpansionHubMotor BL, ExpansionHubMotor BR) {
        frontLeft = FL;
        frontRight = FR;
        backLeft = BL;
        backRight = BR;

        motors = new ExpansionHubMotor[]{frontLeft, frontRight, backLeft, backRight};
        frontRight.setDirection(DcMotorSimple.Direction.REVERSE);
        backRight.setDirection(DcMotorSimple.Direction.REVERSE);
        for (ExpansionHubMotor motor : motors) {
            motor.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
            motor.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
        }
    }

    public static void stopMovement() {
        movementX = 0;
        movementY = 0;
        movementTurn = 0;
    }

    public void update() {
        // remember that motors on the right side are flipped (or maybe left i forgot)
        // https://www.vexforum.com/uploads/default/original/2X/c/c7963dec531b7e893ddcf91c4fc9159928c4b56c.jpeg
        double rawFrontLeft = -movementY + movementX + movementTurn;
        double rawBackLeft = -movementY - movementX + movementTurn;
        double rawFrontRight = -movementY + movementX - movementTurn;
        double rawBackRight = -movementY - movementX - movementTurn;
        double[] powers = {rawFrontLeft, rawFrontRight, rawBackLeft, rawBackRight};

        // if its over 1 then lower everything by a similar ratio to keep the same profile
        double biggestAbsPower = Math.abs(rawFrontLeft);
        for (double power : powers) {
            if (Math.abs(power) > biggestAbsPower)
                biggestAbsPower = Math.abs(power);
        }

        // divide everything to <= 1
        if (biggestAbsPower > 1) {
            for (int i = 0; i < powers.length; i++)
                powers[i] = powers[i] / biggestAbsPower;
        }

        for (int i = 0; i < powers.length; i++) {
            motors[i].setPower(powers[i]);
        }
    }

    @Override
    public String toString() {
        return "movementX: " + movementX + " movementY: " + movementY + " movementTurn: " + movementTurn;
    }

    @Override
    public void turnOn() {

    }

    @Override
    public void turnOff() {

    }

    @Override
    public void reverse() {

    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/hardware/DriveTrain.kt
package org.firstinspires.ftc.teamcode.hardware

import com.qualcomm.robotcore.hardware.DcMotor
import org.firstinspires.ftc.robotcore.external.Telemetry
import org.firstinspires.ftc.teamcode.util.Angle
import org.firstinspires.ftc.teamcode.util.Point
import org.firstinspires.ftc.teamcode.util.Pose
import org.openftc.revextensions2.ExpansionHubMotor
import kotlin.math.absoluteValue

class DriveTrain(
    frontLeft: ExpansionHubMotor,
    frontRight: ExpansionHubMotor,
    backLeft: ExpansionHubMotor,
    backRight: ExpansionHubMotor
) : Hardware() {

    var powers: Pose
    private val motors = arrayOf(frontLeft, frontRight, backLeft, backRight)

    override fun update(telemetry: Telemetry) {
        val rawFrontLeft: Double = -powers.y - powers.x + powers.h.raw
        val rawFrontRight: Double = powers.y - powers.x + powers.h.raw
        val rawBackLeft: Double = -powers.y + powers.x + powers.h.raw
        val rawBackRight: Double = powers.y + powers.x + powers.h.raw

        var rawPowers = listOf(rawFrontLeft, rawFrontRight, rawBackLeft, rawBackRight)
        val max: Double = rawPowers.map { it.absoluteValue }.maxOrNull()!!
        if (max > 1.0) rawPowers = rawPowers.map { it / max }

        motors.forEachIndexed { i, it -> it.power = rawPowers[i] }

//        val dp = LinkedHashMap<String, String>()
//        dp["power vectors"] = powers.toRawString
//        dp["powers"] = rawPowers.toString()
//        return dp
    }

    fun setZeroPowers() {
        powers = Pose(Point(), Angle(0.0, Angle.Unit.RAW))
    }

    init {
        for (m in motors) {
            m.mode = DcMotor.RunMode.RUN_WITHOUT_ENCODER
            m.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        }
        powers = Pose(Point(), Angle(0.0, Angle.Unit.RAW))
    }
}
